plugins/inputs/cloudwatch/cloudwatch.go

```diff
@@ -149,11 +145,11 @@ func (c *CloudWatch) Gather(acc telegraf.Accumulator) error {
 			} else {
 				allMetrics, err := c.fetchNamespaceMetrics()
 				if err != nil {
-					return err
+					return nil, err
 				}
 				for _, name := range m.MetricNames {
 					for _, metric := range allMetrics {
-						if isSelected(metric, m.Dimensions) {
+						if name == *metric.MetricName && isSelected(metric, m.Dimensions) {
```

Comment:
The name logic check should just be folded into the `isSelected` method.
So, `isSelected(name, metric, m.Dimensions)`